### PR TITLE
[ml] `TorchTrainer` bug fixes + GPU test

### LINKED ISSUE #23297: [ml] TorchTrainer docstring doesn't work
https://github.com/ray-project/ray/blob/04ab27dcbf61226d5b53ef3f5375a98308d6a460/python/ray/ml/train/integrations/torch/torch_trainer.py#L136-L138

Doesn't show how to pass in train_loop.

### LINKED ISSUE #23300: [ml] Torch train example should also show GPU
User has to infer how to use the Scaling Config. 

We can make this easier by just providing a comment in the docstring

### LINKED ISSUE #23302: [ml] Fix 'num_gpus_per_worker'
trainer = TorchTrainer(
   train_loop_per_worker=train_func,
   scaling_config={"num_workers": 1, "use_gpu": True},
)
trainer.fit()

AttributeError: 'dict' object has no attribute 'num_gpus_per_worker'

### LINKED ISSUE #23308: [ml][Improvement] Make sure train.report and train.save_checkpoint is called on all the actors → this error message needs to be corrected

### PR DESCRIPTION
**Pull request by amogkam**
Closes #23302 
Closes https://github.com/ray-project/ray/issues/23300
Closes https://github.com/ray-project/ray/issues/23297
Closes https://github.com/ray-project/ray/issues/23308

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by richardliaw** [comment]
Looks good, what do you think about just closing a bunch of the issues together in one PR?

**Comment by amogkam** [comment]
What other issues do you think are relevant here? It seems like the other issues either require some more effort to fix or involve different components.

**Comment by amogkam** [comment]
Ok added a couple more. I think for everything else, we should do in separate PRs.

**Review by matthewdeng — APPROVED** [review]

File: python/ray/ml/train/integrations/torch/torch_trainer.py [hunk 1]
```diff
@@ -134,7 +134,12 @@ def train_loop_for_worker():
                     train.save_checkpoint(model=model.state_dict())
 
             train_dataset = ray.data.from_items([1, 2, 3])
-            trainer = TorchTrainer(scaling_config={"num_workers": 3},
+            scaling_config = {"num_workers": 3}
+            # If using GPUs, use the below scaling config instead.
+            # scaling_config = {"num_workers": 3, "use_gpu" True}
```
Comment (ericl) [inline python/ray/ml/train/integrations/torch/torch_trainer.py:139]:
```suggestion
            # scaling_config = {"num_workers": 3, "use_gpu": True}
```

**Review by ericl — APPROVED** [review]
LGTM sans typo

**Comment by richardliaw** [comment]
tests fail?

**Pull request merged**